### PR TITLE
Add FunctionCall expression rewriter to presto.verifier

### PR DESCRIPTION
**Pull request by gggrace14**
This change creates a FunctionCallRewriter that overrides the base class AstVisitor,
substituting the FunctionCall node and reconstructing the AST of the source query.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/DefaultTreeRewriter.java [hunk 1]
```diff
@@ -0,0 +1,729 @@
...
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * A default implementation of {@link AstVisitor} that reconstructs a node if any of its children is reconstructed.
+ * Expression node reconstruction is not supported and left to the users.
+ */
+public class DefaultTreeRewriter<C>
```
Comment (agrawaldevesh) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/DefaultTreeRewriter.java:87]: If this class is copied and forked, can you link to the original and mention the motivation and change ?

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/FunctionCallRewriter.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(name, originalArgumentIndices);
+        }
+    }
+
+    private static final Pattern FUNCTION_CALL_SUBSTITUTION_PATTERN = Pattern.compile("/(\\w+)\\(([\\w|,]+)\\)/(\\w+)\\(([\\w|,]+)\\)/");
```
Comment (agrawaldevesh) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/FunctionCallRewriter.java:69]: Can you add a comment on what this regex matches ?

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java [hunk 3]
```diff
@@ -159,6 +172,9 @@ public QueryObjectBundle rewriteQuery(@Language("SQL") String query, ClusterType
             ResultSetMetaData metadata = getResultMetadata((Query) statement);
             List<Identifier> columnAliases = generateStorageColumnAliases(metadata);
             Query rewrite = rewriteNonStorableColumns((Query) statement, metadata);
+            if (functionCallRewriter.isPresent()) {
```
Comment (agrawaldevesh) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java:175]: Why is the function call rewrite only done for select queries ?

File: presto-verifier/src/test/java/com/facebook/presto/verifier/rewrite/TestQueryRewriter.java [hunk 4]
```diff
@@ -294,6 +298,160 @@ public void testRewriteNonStorableStructuredTypes()
...
+
+        assertTrue(FunctionCallRewriter.constructFunctionCallSubstituteMap("").isEmpty());
+        assertTrue(FunctionCallRewriter.constructFunctionCallSubstituteMap("/func1(a,b)//").isEmpty());
+    }
+
+    @Test
+    public void testRewriteFunctionCalls()
+    {
+        QueryRewriter queryRewriter = getQueryRewriter(Optional.of("/approx_distinct(x)/count(x)/," +
+                "/approx_percentile(x,_)/avg(x)/," +
```
Comment (agrawaldevesh) [inline presto-verifier/src/test/java/com/facebook/presto/verifier/rewrite/TestQueryRewriter.java:326]: I am wary of dropping arguments like this: we may be changing what gets read transitively and maybe hiding any bugs. 

Particularly as we test newer file formats like alpha with velox. 

Is there a way to continue projecting the existing arguments somehow ? Like avg plus some function of the second argument ?

File: presto-verifier/src/test/java/com/facebook/presto/verifier/rewrite/TestQueryRewriter.java [hunk 5]
```diff
@@ -294,6 +298,160 @@ public void testRewriteNonStorableStructuredTypes()
...
+                "/arbitrary(x)/min(x)/," +
+                "/map_agg(x,)/max(x)/," +
+                "/first_value(x)/min(x)/," +
+                "/map_union(x)/max(x)/," +
+                "/array_agg(x)/max(x)/"));
+
+        // Test rewriting window functions
+        assertCreateTableAs(
```
Comment (agrawaldevesh) [inline presto-verifier/src/test/java/com/facebook/presto/verifier/rewrite/TestQueryRewriter.java:334]: @mbasmanova had made a comment to test some failure scenarios like where the substitute is invalid. Where are those to see that the rewrite is to blame ?

File: presto-verifier/src/test/java/com/facebook/presto/verifier/rewrite/TestQueryRewriter.java (hunk 4 above)
Comment (spershin) [reply presto-verifier/src/test/java/com/facebook/presto/verifier/rewrite/TestQueryRewriter.java:326]: There will be some limitations on 'function replace' approach, of course.
In this particular case approx_percentile() most of the time has the rest of the arguments as constants.
The case with weights is exception.
Having a generic replacement as avg(x + func1(y)) can have a problem of type incompatibility.
We could try and se if this will work all the time:
"/approx_percentile(x,y,_)/avg(x + y)/,"
"/approx_percentile(x)/avg(x)/,"

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java (hunk 3 above)
Comment (spershin) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java:175]: I'm curious is where does it say it is only for "select" queries?

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java [hunk 6]
```diff
@@ -95,19 +95,32 @@ public class QueryRewriter
...
+            Map<ClusterType, QualifiedName> tablePrefixes,
+            Map<ClusterType, List<Property>> tableProperties,
+            Optional<String> nonDeterministicFunctionSubstitutes)
     {
         this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
         this.typeManager = requireNonNull(typeManager, "typeManager is null");
         this.prestoAction = requireNonNull(prestoAction, "prestoAction is null");
         this.prefixes = ImmutableMap.copyOf(tablePrefixes);
         this.tableProperties = ImmutableMap.copyOf(tableProperties);
+        this.functionCallRewriter = requireNonNull(nonDeterministicFunctionSubstitutes, "nonDeterministicFunctionSubstitutes is null").map(FunctionCallRewriter::getInstance);
```
Comment (spershin) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java:123]: I don't understand this:
1. nonDeterministicFunctionSubstitutes is Optional<String>, meaning it can be none.
2. We use requireNonNull, meaning is cannot be 'none'?

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/FunctionCallRewriter.java (hunk 2 above)
Comment (spershin) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/FunctionCallRewriter.java:69]: +1
And one or two examples. It takes me a significant effort every time to understand any kind of regex - I don't work with them too often.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateMaterializedView.java [hunk 7]
```diff
@@ -31,6 +31,16 @@ public class CreateMaterializedView
...
+    public CreateMaterializedView(QualifiedName name, Query query, boolean notExists, List<Property> properties, Optional<String> comment)
+    {
+        super(Optional.empty());
+        this.name = requireNonNull(name, "name is null");
+        this.query = requireNonNull(query, "query is null");
+        this.notExists = notExists;
+        this.properties = ImmutableList.copyOf(requireNonNull(properties, "properties is null"));
+        this.comment = requireNonNull(comment, "comment is null");
+    }
+
```
Comment (spershin) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateMaterializedView.java:43]: Why is this needed in this PR?

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/FunctionCallRewriter.java [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            return map.build();
+        }
+
+        Matcher matcher = FUNCTION_CALL_SUBSTITUTION_PATTERN.matcher(functionCallSubstitutes);
+
+        while (matcher.find()) {
+            String originalName = matcher.group(1);
+            List<String> originalArgumentList = ImmutableList.copyOf(matcher.group(2).split(","));
+            String substituteName = matcher.group(3);
+            List<String> substituteArgumentList = ImmutableList.copyOf(matcher.group(4).split(","));
```
Comment (spershin) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/FunctionCallRewriter.java:100]: This ".group(N)", is N a 1-based index?

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java (hunk 3 above)
Comment (agrawaldevesh) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java:175]: The above two if clauses are for insert and ctas -- so this one is for select queries. We should be rewriting the above two as well I think. Esp for batch workload which has a lot of inserts.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateMaterializedView.java (hunk 7 above)
Comment (gggrace14) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateMaterializedView.java:43]: To provide a constructor without the arg Optional<NodeLocation> location. Almost all subclasses of Node has similar constructor overloading, ex., CreateTableAsSelect, and it is missing for CreateMaterializedView. This is needed by an Ast rewriter DefaultTreeRewriter, since the old NodeLocation is not valid anymore after reconstructing.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/FunctionCallRewriter.java (hunk 2 above)
Comment (gggrace14) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/FunctionCallRewriter.java:69]: Split the pattern definition, and hope it makes it a bit easier to read.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/FunctionCallRewriter.java (hunk 8 above)
Comment (gggrace14) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/FunctionCallRewriter.java:100]: .group(1) refers to the group wrapped by the first paranthesis. .group(0) always refers to the entire match. Ex., for a(bc), group(0) returns abc, and group(1) return bc, if matched.

(Also got my memory refreshed by reading the java doc another time)

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java (hunk 6 above)
Comment (gggrace14) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java:123]: For Java,
1) Optional<String> nonDeterministicFunctionSubstitutes = Optional.empty();
2) Optional<String> nonDeterministicFunctionSubstitutes = null;

are different. 2) cannot pass requireNonNull() while 1) can.

Separately, this line calls Optional.map(), and if FunctionCallRewriter.getInstance(nonDeterministicFunctionSubstitutes.get()) returns null, Optional.map() returns Optional.empty().

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java (hunk 3 above)
Comment (gggrace14) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java:175]: The if-flow here now is
```
if (statement instanceof CreateTableAsSelect) {
}
if (statement instanceof Insert) {
}
if (statement instanceof Query) {
    rewrite = (Query) functionCallRewriter.get().process(rewrite);
}
if (statement instanceof CreateView) {
}
```

I will add `functionCallRewriter.get().process()` to the rest in my 2nd PR, https://github.com/prestodb/presto/pull/18095. This PR adds it to the Query branch, only for the sake of unit tests that utilizes existing setup.

This change is split into 2 PRs,
This https://github.com/prestodb/presto/pull/20923 to add the FunctionCallRewriter class.
https://github.com/prestodb/presto/pull/18095 to invoke the FunctionCallRewriter class in the QueryRewriter class  as well as configs.

File: presto-verifier/src/test/java/com/facebook/presto/verifier/rewrite/TestQueryRewriter.java (hunk 5 above)
Comment (gggrace14) [reply presto-verifier/src/test/java/com/facebook/presto/verifier/rewrite/TestQueryRewriter.java:334]: For invalid substitution, I include the unit test testConstructMakeFunctionCallSubstituteMap() in this PR to validate the illegal substitution specification string, ex., /func1(a,b)//.

For validating illegal function signature or type signature, like /map_agg(x,)/max(x)/, I will handle this in https://github.com/prestodb/presto/pull/18095.
(The best I figured we can do in Verifier without access to HiveMetadata is to actually run the query after rewritten, fail and provide meaningful error info.)

**Review by spershin — APPROVED** [review]
This looks good to me, but I'm not a committer, so someone else would have to approve this change as well.
cc @mbasmanova 

If there are minor things, we can add them in the next PR or even after, when we start running this for Presto Native and learn some lesson on practice.

**Review by agrawaldevesh — APPROVED** [review]

**Review by mbasmanova — APPROVED** [review]
